Movable insulation logic refactoring

xuanluo113 opened the pull request:
Pull request overview
---------------------
The branch improves the logic to calculate movable insulation properties. The ext and int heat transfer coefficients and absorptances only need to calculated once at the beginning of the heat balance surface manager when any movable insulation exists. The subroutines of calculating these properties are transformed from external function calls to loop-wise initializations. The functions are moved to HeatBalanceSurfaceManager and the original movable insulation related files are removed.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] ~If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix~
 - [ ] ~If feature, test running new feature, try creative ways to break it~
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] ~Verify IDF naming conventions and styles, memos and notes and defaults~
 - [ ] ~If new idf included, locally check the err file and other outputs~

xuanluo113 committed:
movable insulation simplified

xuanluo113 committed:
unit tests

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into movable-insul

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into movable-insul

xuanluo113 committed:
initialization

xuanluo113 committed:
merge

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into movable-insul

xuanluo113 committed:
fix initialization

xuanluo113 committed:
merge

xuanluo113 committed:
merge

xuanluo113 committed:
merge

xuanluo113 commented:
@mjwitte This branch is ready for review. CI and performance tests look green. Thanks very much!

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into movable-insul

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into movable-insul

mjwitte commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -1467,11 +1475,17 @@ void AllocateSurfaceHeatBalArrays(EnergyPlusData &state)
     state.dataSurface->SurfGndSolarInc.dimension(state.dataSurface->TotSurfaces, 0);
 
+    // allocate movable insulation arrays
+    state.dataSurface->SurfMovInsulIntPresent.dimension(state.dataSurface->TotSurfaces, false);
+    state.dataSurface->SurfMovInsulIntPresentPrevTS.dimension(state.dataSurface->TotSurfaces, false);
+    state.dataSurface->SurfMovInsulHExt.dimension(state.dataSurface->TotSurfaces, 0.0);
+    state.dataSurface->SurfMovInsulAbsExt.dimension(state.dataSurface->TotSurfaces, 0.0);
+    state.dataSurface->SurfMovInsulHInt.dimension(state.dataSurface->TotSurfaces, 0.0);
+    state.dataSurface->SurfMovInsulAbsInt.dimension(state.dataSurface->TotSurfaces, 0.0);
```
Seems excessive to allocate all of these arrays when typically there will be no surfaces with movable insulation, and at most only a few of the surfaces in the entire model will have it.

amirroth replied:
The alternative would be to ... have a single array that indexes to a set of smaller arrays (or even one arrays of structures) sized to fit only the number of surfaces that have `MovableInsulation`? :thinking: How do we feel about that design pattern in general? How broadly should we apply it? I have been thinking about whether we should do something like this for all of the Window specific arrays. :thinking: :thinking:

xuanluo113 replied:
Five HB arrays are added including `SurfAbsSolarExt`, `SurfAbsThermalExt`, `SurfAbsSolarInt`, `SurfAbsThermalInt`, SurfRoughnessExt`. They are initialized with material properties one time at the beginning of `InitSurfaceHeatBalance`, and are set to movable insulation properties (and set back) at run time when `AnyMovableInsulation`.

xuanluo113 replied:
`SurfMovInsulExtPresent`, `SurfMovInsulIntPresent`, `SurfMovInsulIntPresentPrevTS`, `SurfMovInsulHExt`, `SurfMovInsulHInt` are allocated only when `AnyMovableInsulation`, and whenever they are used (e.g. before checking if `SurfMovInsulExtPresent`), `AnyMovableInsulation` is checked first.

mjwitte commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -2292,6 +2306,55 @@ void InitThermalAndFluxHistories(EnergyPlusData &state)
+void EvalOutsideMovableInsulation(EnergyPlusData &state)
+{
+    // This subroutine determines whether or not outside movable insulation on opaque surfaces is present at the current time.
+    // The SurfNum is passed in and then the rest of the parameters are set if movable insulation is present.
```
This comment is not relevant here.

xuanluo113 replied:
Deleted.

mjwitte commented on src/EnergyPlus/HeatBalanceManager.cc:
```diff
@@ -5940,7 +5942,7 @@ namespace HeatBalanceManager {
         // exchange algorithm knows whether there has been a change in interior movable insulation or not.
         for (SurfNum = 1; SurfNum <= state.dataSurface->TotSurfaces; ++SurfNum) {
-            state.dataSurface->Surface(SurfNum).MovInsulIntPresentPrevTS = state.dataSurface->Surface(SurfNum).MovInsulIntPresent;
+            state.dataSurface->SurfMovInsulIntPresentPrevTS(SurfNum) = state.dataSurface->SurfMovInsulIntPresent(SurfNum);
```
This loop can be skipped if `!AnyMovableInsulation`.

xuanluo113 replied:
Flag added.

mjwitte commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -2292,6 +2306,55 @@ void InitThermalAndFluxHistories(EnergyPlusData &state)
+{
+    // This subroutine determines whether or not outside movable insulation on opaque surfaces is present at the current time.
+    // The SurfNum is passed in and then the rest of the parameters are set if movable insulation is present.
+    for (int zoneNum = 1; zoneNum <= state.dataGlobal->NumOfZones; ++zoneNum) {
```
Even if we keep the full arrays for movable insulation, it would be useful to build a list of surfaces with movable insulation and simply loop over that list here instead of looping over every opaque surface.

xuanluo113 replied:
A vector `SurfMovInsulIndexList` is initialized at the input stage and the loops are reduced.

mjwitte commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -2854,19 +2916,15 @@ void InitSolarHeatGains(EnergyPlusData &state)
-
-                    if (RoughIndexMovInsul <= 0) { // No movable insulation present
-
+                    if (state.dataSurface->SurfMovInsulHExt(SurfNum) > 0) {
+                        AbsExt = state.dataSurface->SurfMovInsulAbsExt(SurfNum);
+                    } else { // No movable insulation present
                         AbsExt = state.dataMaterial->Material(state.dataConstruction->Construct(ConstrNum).LayerPoint(1)).AbsorpSolar;
                     }
```
This if construct could be eliminated if we had a `SurfAbsExt` array that is intialized to the base construction `AbsorpSolar` and modified or reset as needed in `EvalOutsideMoveableInsulation`.

amirroth replied:
Just factoring this code into a separate loop and guarding it with `if (MovableInsulationAnywhereInTheModel)` will speed up models that have no `MovableInsulation` in them. A "push" implementation like the one @mjwitte suggests will also speed up models that do have `MovableInsulation` in them, although do we really care about such models right now?

xuanluo113 replied:
Five HB arrays are added including SurfAbsSolarExt, SurfAbsThermalExt, SurfAbsSolarInt, SurfAbsThermalInt, SurfRoughnessExt. They are initialized with material properties one time at the beginning of InitSurfaceHeatBalance, and are set to movable insulation properties (and set back) at run time when AnyMovableInsulation`.

mjwitte commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -3567,22 +3615,18 @@ void InitIntSolarDistribution(EnergyPlusData &state)
+
+            if (state.dataSurface->SurfMovInsulHInt(SurfNum) > 0) {
+                AbsIntSurf = state.dataSurface->SurfMovInsulAbsInt(SurfNum);
+            }
```
Similar comment here. `SurfIntAbs` and `SurfIntAbsVis` would eliminate this if and the repeated construction lookups above.

xuanluo113 replied:
Like explained in the previous comment, `AbsorpSolar` and `AbsorpThermal` are saved to runtime HB arrays. I removed `SurfMovInsulAbsInt` array here and added `SurfAbsSolarInt` instead.

xuanluo113 replied:
In [develop branch](https://github.com/NREL/EnergyPlus/blob/b622e27ca3c96566d6bbc8c6429ae971f880d445/src/EnergyPlus/HeatBalanceSurfaceManager.cc#L3570), `AbsIntSurfVis` is set as `Construct(ConstrNum).InsideAbsorpSolar` instead of `InsideAbsorpVis`, and is not reset when movable insulation is on. Although the comment here says "Inside opaque surface visible absorptance to fix CR 8695 change to this = Construct(ConstrNum)%InsideAbsorpVis". I wonder if this is a bug or it's intended to be fixed like this. @mjwitte Could you help take a look?

mjwitte commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -3567,22 +3615,18 @@ void InitIntSolarDistribution(EnergyPlusData &state)
-            HMovInsul = 0.0;
-            if (Surface(SurfNum).MaterialMovInsulExt > 0) EvalOutsideMovableInsulation(state, SurfNum, HMovInsul, RoughIndexMovInsul, AbsExt);
-            if (HMovInsul > 0) { // Movable outside insulation in place
+            if (state.dataSurface->SurfMovInsulHExt(SurfNum) > 0) { // Movable outside insulation in place
```
Because exterior movable insulation doesn't have a `Present` variable like inside movable insulation, `SurfMovInsulHExt >0` is being used to test for presence. In the old `EvalOutsideMovableInsulation` function, this was reset to zero when `MovInsulSchedVal <= 0.0` but I don't see that in the new version of this function.

amirroth replied:
We should probably implement inside and outside `MovableInsulation` in similar ways, no? Consistency is the hobgoblin of little minds. And of Amir.

xuanluo113 replied:
I added `SurfMovInsulExtPresent` and used the boolean arrays in checking presence for both inside and outside surface.

mjwitte commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -3979,13 +4020,8 @@ void ComputeIntThermalAbsorpFactors(EnergyPlusData &state)
-                if (HMovInsul > 0.0)
-                    state.dataHeatBal->ITABSF(SurfNum) =
-                        state.dataMaterial->Material(Surface(SurfNum).MaterialMovInsulInt).AbsorpThermal; // Movable inside insulation present
+            if (state.dataSurface->SurfMovInsulHInt(SurfNum) > 0.0) {
+                state.dataHeatBal->ITABSF(SurfNum) = state.dataMaterial->Material(Surface(SurfNum).MaterialMovInsulInt).AbsorpThermal; // Movable inside insulation present
             }
```
Seems that `EvalInsideMovableInsulation` could manage setting `ITABSF`.  Seems this entire loop could be eliminated then. But what about EMS construction changes.

For your todo list: `ComputeIntThermalAbsorpFactors` and `ComputeIntSWAbsorpFactors` appear to be completely recomputed every timestep? Seems these function are ripe for skipping a zone if nothing has changed with window shade status or movable insulation? Not suggesting you try that here.

xuanluo113 replied:
Runtime `SurfAbsThermalExt` is used instead here and initialized with `state.dataSurface->SurfActiveConstruction` at the beginning of simulation or when any EMS override is on.

mjwitte commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -5667,14 +5701,10 @@ void ReportSurfaceHeatBalance(EnergyPlusData &state)
-            if (GetCurrentScheduleValue(state, Surface(SurfNum).SchedMovInsulInt) > 0.0) {
-                state.dataHeatBalSurf->TempSurfInMovInsRep(SurfNum) = state.dataHeatBalSurf->TempSurfInTmp(SurfNum);
-            }
+    for (int SurfNum = 1; SurfNum <= state.dataSurface->TotSurfaces; ++SurfNum) {
```
Another spot for a list-driven update.

xuanluo113 replied:
Used the index list instead.

mjwitte commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -6136,13 +6165,11 @@ void CalcHeatBalanceOutsideSurf(EnergyPlusData &state,
-                        if (HMovInsul > 0)
-                            AbsThermSurf = state.dataMaterial->Material(Surface(SurfNum).MaterialMovInsulExt)
-                                               .AbsorpThermal; // Movable outside insulation present
+                    if (state.dataSurface->SurfMovInsulHExt(SurfNum) > 0) {
```
Another place where arrays of surface roughness and absthermal could eliminate this if and repeated lookups from `dataConstruction`.

mjwitte reviewed (commented):
@xuanluo113 This is a good start. I think there's room for more improvement. See comments below.

mjwitte commented:
And there are clang formatting check errors - need to apply formatting to all of the changed files.

xuanluo113 committed:
address comments

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into movable-insul

xuanluo113 committed:
addressing comments

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into movable-insul

xuanluo113 committed:
address comments fix

xuanluo113 committed:
clang format

xuanluo113 committed:
Revert "clang format"

This reverts commit 00c49918133fcdd62a71d6abba6d6350c477bb08.

xuanluo113 committed:
clang-format

xuanluo113 committed:
fix unit test and EMS

xuanluo113 committed:
clang-format

xuanluo113 commented:
@mjwitte Thanks for the review. I've finished addressing most of the comments as replied, except for two items to work on here or later.

1. The `AbsIntSurfVis` bug (?) as above.
2. `dataHeatBal->ITABSF` seems to have the same physical meaning as the newly added `SurfAbsThermalExt` array, while `ITABSF` is used in window surfaces and `SurfAbsThermalExt` in opaque surfaces. I intend to remove `ITABSF`, but this can be done along with the todo item listed above - reducing `ComputeIntThermalAbsorpFactors` and `ComputeIntSWAbsorpFactors`?

xuanluo113 committed:
clang-format

mjwitte commented on src/EnergyPlus/DataGlobals.hh:
```diff
@@ -141,6 +141,7 @@ struct DataGlobal : BaseGlobalStruct
     bool RedoSizesHVACSimulation = false;           // doing kick off simulation for redoing sizes as part of sizing
     bool FinalSizingHVACSizingSimIteration = false; // when doing HVAC sizing Simulation
     bool AnyEnergyManagementSystemInModel = false;  // true if there is any EMS or Erl in model.  otherwise false
+    bool SurfPropOverridesPresent = false;          // true if there is any EMS or Erl in model.  otherwise false
```
This endline comment should be more specific. And probably should be named `AnySurfPropOverridesInModel` to be consistent with the others? This can be be a TODO for later if you want.

xuanluo113 replied:
I've fixed this comment and naming. Waiting for the local tests to finish for removing `ConstrNum > 0`. Will make a push later.

mjwitte commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -395,13 +394,35 @@ void InitSurfaceHeatBalance(EnergyPlusData &state)
+            int const lastSurf = state.dataHeatBal->Zone(zoneNum).HTSurfaceLast;
+            for (int SurfNum = firstSurf; SurfNum <= lastSurf; ++SurfNum) {
+                int ConstrNum = state.dataSurface->SurfActiveConstruction(SurfNum); // SurfActiveConstruction set above
+                if (ConstrNum > 0) {
```
Is it necessary to check for `ConstrNum>0`? All heat transfer surfaces should have a valid construction.

mjwitte commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -3566,23 +3622,18 @@ void InitIntSolarDistribution(EnergyPlusData &state)
 
-            Real64 AbsIntSurf = state.dataConstruction->Construct(ConstrNum).InsideAbsorpSolar; // Inside opaque surface solar absorptance
+            Real64 AbsIntSurf = state.dataHeatBalSurf->SurfAbsSolarInt(SurfNum);
+            // TODO - AbsIntSurfVis = InsideAbsorpSolar instead of InsideAbsorpVis?
```
If I'm following correctly, `AbsIntSurfVis` is only used for reporting, not in the actual surface heat gain calculations. Per the comments in [CR8695](https://github.com/NREL/EnergyPlus/issues/3468) it sounds like these needs to remain AbsSolar so that the reporting is consistent with the actual heat balance. But it's certainly worth looking at. It might be possible to fix that (later).

mjwitte reviewed (approved):
@xuanluo113 This is so much better with much cleaner logic. A few minor comments below.
I haven't looked at the ITABSF question yet.
I'll wait for CI to finish before merging. Or if you want to address the SurfPropOverridesPresent comment in this PR, let me know.

xuanluo113 committed:
fix a comment

mjwitte commented:
@xuanluo113 Looks good. Merging.

jcyuan2020 commented:
This PR gets an extra credit for fixing a newly reported issue #8758.
